File: matrix-app-front/libs/matrix-thales-d1/android/src/main/java/com/com.thalesgroup.gemalto.d1.validation/helper/D1Helper.java
package com.thalesgroup.gemalto.d1.validation.helper;

import android.app.Activity;
import android.content.Intent;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.firebase.messaging.FirebaseMessaging;
import com.thalesgroup.gemalto.d1.CardPINUI;
import com.thalesgroup.gemalto.d1.ConfigParams;
import com.thalesgroup.gemalto.d1.D1Exception;
import com.thalesgroup.gemalto.d1.D1Params;
import com.thalesgroup.gemalto.d1.D1Task;
import com.thalesgroup.gemalto.d1.EntryUI;
import com.thalesgroup.gemalto.d1.PINDisplayTextView;
import com.thalesgroup.gemalto.d1.SecureEditText;
import com.thalesgroup.gemalto.d1.card.CardAction;
import com.thalesgroup.gemalto.d1.card.CardActivationMethod;
import com.thalesgroup.gemalto.d1.card.CardDataChangedListener;
import com.thalesgroup.gemalto.d1.card.CardDetails;
import com.thalesgroup.gemalto.d1.card.CardDigitizationState;
import com.thalesgroup.gemalto.d1.card.CardMetadata;
import com.thalesgroup.gemalto.d1.card.DigitalCard;
import com.thalesgroup.gemalto.d1.card.OEMPayType;
import com.thalesgroup.gemalto.d1.validation.util.HexUtil;

import java.util.List;

/**
 * Helper class for D1 SDK.
 */
public class D1Helper {


    private static final D1Helper INSTANCE = new D1Helper();

    private D1Task mD1Task;
    private Boolean mSdkIsConfigured = false;

    private D1Helper() {
        // private constructor
    }

    /**
     * Retrieves the singleton instance of {@code D1Helper}.
     *
     * @return Singleton instance of {@code D1Helper}.
     */
    @NonNull
    public static D1Helper getInstance() {
        return INSTANCE;
    }

    /**
     * Configures the D1 SDK.
     *
     * @param consumerId         Consumer ID.
     * @param activity           Activity.
     * @param callback           Callback.
     */
    
    public void configure(
                        @NonNull final String serviceUrl,
                        @NonNull final String issuerId,
                        @NonNull final String exponent,
                        @NonNull final String modulus,
                        @NonNull final String digitalCardUrl,
                        @NonNull final String consumerId,
                        @NonNull final Activity activity,
                        @NonNull final D1Task.ConfigCallback<Void> callback) {
        // D1Core config.
        mD1Task = new D1Task.Builder().setContext(activity.getApplicationContext()).setD1ServiceURL(serviceUrl)
                .setIssuerID(issuerId)
                .setD1ServiceRSAExponent(HexUtil.hexStringToByteArray(exponent))
                .setD1ServiceRSAModulus(HexUtil.hexStringToByteArray(modulus))
                .setDigitalCardURL(digitalCardUrl).build();

        final D1Params coreConfig = ConfigParams.buildConfigCore(consumerId);
        final D1Params cardConfig = ConfigParams.buildConfigCard(activity, OEMPayType.GOOGLE_PAY, null);


        mD1Task.configure(new D1Task.ConfigCallback<Void>() {
            @Override
            public void onSuccess(final Void unused) {
                synchronized (INSTANCE) {
                    mSdkIsConfigured = true;
                }

                getCurrentPushToken();
                callback.onSuccess(unused);
            }

            @Override
            public void onError(@NonNull final List<D1Exception> list) {
                callback.onError(list);
            }
        }, coreConfig,cardConfig);
    }

    /**
     * Logs in.
     *
     * @param issuerToken Issuer token.
     * @param callback    Callback.
     */
    public void login(@NonNull final byte[] issuerToken, @NonNull final D1Task.Callback<Void> callback) {
        getD1Task().login(issuerToken, callback);
    }

    /**
     * Logs out.
     *
     * @param callback Callback.
     */
    public void logout(@NonNull final D1Task.Callback<Void> callback) {
        getD1Task().logout(callback);
    }

    public void handleResult(final int requestCode, final int resultCode, @Nullable final Intent data) {
        getD1Task().handleCardResult(requestCode, resultCode, data);
    }

    /**
     * Retrieves the card meta data.
     *
     * @param cardId   Card ID.
     * @param callback Callback.
     */
    public void getCardMetadata(@NonNull final String cardId, @NonNull final D1Task.Callback<CardMetadata> callback) {
        getD1Task().getCardMetadata(cardId, callback);
    }

    /**
     * Retrieves the card meta data.
     *
     * @param cardId   Card ID.
     * @param callback Callback.
     */
    public void getCardDetails(@NonNull final String cardId, @NonNull final D1Task.Callback<CardDetails> callback) {
        getD1Task().getCardDetails(cardId, callback);
    }

    /**
     * Checks if card is digitized.
     *
     * @param cardId   Card ID.
     * @param callback Callback.
     */
    public void getCardDigitizationState(@NonNull final String cardId,
                                         @NonNull final D1Task.Callback<CardDigitizationState> callback) {

        getD1Task().getD1PushWallet().getCardDigitizationState(cardId, OEMPayType.GOOGLE_PAY, callback);

    }

    /**
     * Digitizes the card.
     *
     * @param cardId   Card ID.
     * @param callback Callback.
     */
    public void digitizeCard(@NonNull final String cardId, Activity activity, @NonNull final D1Task.Callback<Object> callback) {
        getD1Task().getD1PushWallet().addDigitalCardToOEM(cardId, OEMPayType.GOOGLE_PAY, activity, callback);
    }

    /**
     * Activates the digital card.
     *
     * @param cardId   Card ID.
     * @param callback Callback.
     */
    public void activateDigitalCard(@NonNull final String cardId, @NonNull final D1Task.Callback<Void> callback) {
        getD1Task().activateDigitalCard(cardId, callback);
    }

    /**
     * Activates the physical card.
     *
     * @param cardId         Card ID.
     * @param secureTextEdit Secure text edit, where the code entered.
     * @param callback       Callback.
     */
    public void activatePhysicalCard(@NonNull final String cardId,
                                     @NonNull final SecureEditText secureTextEdit,
                                     @NonNull final D1Task.Callback<Void> callback) {
        final EntryUI entryUI = new EntryUI(secureTextEdit);
        getD1Task().activatePhysicalCard(cardId, entryUI, callback);
    }

    /**
     * Retrieves the PIN of the physical card.
     *
     * @param cardId             Card ID.
     * @param pinDisplayTextView PINDisplayTextView where the PIN is displayed.
     * @param callback           Callback.
     */
    public void getPhysicalCardPin(@NonNull final String cardId,
                                   @NonNull final PINDisplayTextView pinDisplayTextView,
                                   @NonNull final D1Task.Callback<Void> callback) {
        final CardPINUI cardPINUI = new CardPINUI(pinDisplayTextView);
        getD1Task().displayPhysicalCardPIN(cardId, cardPINUI, callback);
    }

    /**
     * Retrieves card activation method.
     *
     * @param cardId   Card ID.
     * @param callback Callback.
     */
    public void getActivationMethod(@NonNull final String cardId,
                                    @NonNull final D1Task.Callback<CardActivationMethod> callback) {
        getD1Task().getCardActivationMethod(cardId, callback);
    }

    /**
     * Activates the physical card.
     *
     * @param cardId   Card ID.
     * @param callback Callback.
     */

    /**
     * Checks if the D1 SDK is configured.
     *
     * @return {@code True} if configured, else {@code false}.
     */
    public boolean isConfigured() {
        return mD1Task != null;
    }

    /**
     * Retrieves the list of digital cards for D1Push.
     *
     * @param cardId   Card ID.
     * @param callback Callback.
     */
    public void getDigitalCardListD1Push(@NonNull final String cardId,
                                         @NonNull final D1Task.Callback<List<DigitalCard>> callback) {
        getD1Task(). getDigitalCardList(cardId, callback);
    }



    /**
     * Suspends the digital card.
     *
     * @param cardId      Card ID.
     * @param digitalCard Digital card.
     * @param callback    Callback.
     */
    public void suspendDigitalCard(@NonNull final String cardId,
                                   @NonNull final DigitalCard digitalCard,
                                   @NonNull final D1Task.Callback<Boolean> callback) {
        getD1Task().updateDigitalCard(cardId, digitalCard, CardAction.SUSPEND, callback);
    }

    /**
     * Resumes the digital card.
     *
     * @param cardId      Card ID.
     * @param digitalCard Digital card.
     * @param callback    Callback.
     */
    public void resumeDigitalCard(@NonNull final String cardId,
                                  @NonNull final DigitalCard digitalCard,
                                  @NonNull final D1Task.Callback<Boolean> callback) {
        getD1Task().updateDigitalCard(cardId, digitalCard, CardAction.RESUME, callback);
    }

    /**
     * Deletes the digital card.
     *
     * @param cardId      Card ID.
     * @param digitalCard Digital card.
     * @param callback    Callback.
     */
    public void deleteDigitalCard(@NonNull final String cardId,
                                  @NonNull final DigitalCard digitalCard,
                                  @NonNull final D1Task.Callback<Boolean> callback) {
        getD1Task().updateDigitalCard(cardId, digitalCard, CardAction.DELETE, callback);
    }

    /**
     * Updates the push token.
     *
     * @param pushToken Push token to update.
     * @param callback  Callback.
     */
    public void setPushToken(@NonNull final String pushToken, @NonNull final D1Task.Callback<Void> callback) {
        synchronized (INSTANCE) {
            if (mSdkIsConfigured) {
                // SDK already configured.
                mD1Task.updatePushToken(pushToken, callback);
            } else {
                // SDK not yet configured.
                // Notify callback to not block the flow.
                callback.onSuccess(null);
            }
        }
    }



    /**
     * Registers the CardDataChangedListener - D1Push.
     *
     * @param cardDataChangedListener CardDataChangedListener.
     */
    public void registerCardDataChangeListener(@NonNull final CardDataChangedListener cardDataChangedListener) {
        getD1Task().registerCardDataChangedListener(cardDataChangedListener);
    }

    /**
     * Unregisters the CardDataChangedListener - D1Push.
     */
    public void unregisterCardDataChangeListener() {
        getD1Task().unRegisterCardDataChangedListener();
    }




    /**
     * Retrieves the D1 SDK versions.
     *
     * @return D1 SDK versions.
     */
    public String getLibVersions() {
        return D1Task.getSDKVersions().toString();
    }

    /**
     * Retrieves the {@code D1Task} instance.
     *
     * @return {@code D1Task} instance.
     */
    @NonNull
    public D1Task getD1Task() {
        if (mD1Task == null) {
            throw new IllegalStateException("Need to configure D1 SDK first.");
        }

        return mD1Task;
    }

    /**
     * Sets the current push token to D1 SDK.
     */
    private void getCurrentPushToken() {
        FirebaseMessaging.getInstance().getToken().addOnCompleteListener(task -> {
            if (!task.isSuccessful()) {
                return;
            }

            // Get new FCM registration token
            final String token = task.getResult();
            mD1Task.updatePushToken(token, new D1Task.Callback<Void>() {
                @Override
                public void onSuccess(final Void data) {
                    // nothing to do
                }

                @Override
                public void onError(@NonNull final D1Exception exception) {
                    // ignore error
                }
            });
        });
    }

}

File: matrix-app-front/libs/matrix-thales-d1/ios/sdk/D1Helper.swift
/*
 * Copyright © 2022 THALES. All rights reserved.
 */

import Foundation
import D1
import SwiftUI

/// Helper class for D1 SDK.
public class D1Helper {
    
    private var d1Task: D1Task?
    
    private static var sharedD1Helper: D1Helper = {
        return D1Helper()
    }()
    
    private init() {
        
    }

    /// Retrieves the singleton instance of  D1Helper}.
    /// - Returns: Singleton instance of D1Helper.
    public class func shared() -> D1Helper {
        return sharedD1Helper
    }

    public func configure(consumerId: String, components: D1Task.Components, callback: @escaping ([D1Error]?) -> Void) {
        d1Task = components.task()

        let coreConfig: ConfigParams = ConfigParams.coreConfig(consumerID: consumerId)
        let cardConfig: ConfigParams = ConfigParams.cardConfig()
        
        getD1Task().configure([coreConfig, cardConfig], completion: callback)
    }
    
    /// Logs in.
    /// - Parameters:
    ///   - accessToken: Issuer token.
    ///   - callback: Callback.
    public func login(accessToken: String, callback: @escaping (D1Error?) -> Void) {
        if var accessTokenData: Data = accessToken.data(using: .utf8) {
            self.getD1Task().login(&accessTokenData, completion: callback)
        }
    }
    
    /// Logs out.
    /// - Parameter callback: Callback.
    public func logout(callback: @escaping (D1Error?) -> Void) {
        getD1Task().logout(callback)
    }

    /// Retrieves the card meta data.
    /// - Parameters:
    ///   - cardId: Card ID.
    ///   - callback: Callback.
    public func getCardMetadata(cardId: String, callback: @escaping (CardMetadata?, D1Error?) -> Void) {
        self.getD1Task().cardMetadata(cardId, completion: callback)
    }
    
    /**
     * Retrieves the protected card details.
     *
     * @param cardId   Card ID.
     * @param callback Callback.
     */
    
    /// Retrieves the protected card details.
    /// - Parameters:
    ///   - cardId: Card ID.
    ///   - callback: Callback.
    public func getCardDetails(cardId: String, callback: @escaping (CardDetails?, D1Error?) -> Void) -> Void {
        self.getD1Task().cardDetails(cardId, completion: callback)
    }
    
    /// Suspends the card.
    /// - Parameters:
    ///   - cardId: Card ID.
    ///   - callback: Callback.
    public func suspedCard(cardId: String, callback: @escaping (D1Error?) -> Void) {
       // self.getD1Task().updateCard(cardId, cardAction: CardAction.suspend, reason: "Suspend", completion: callback)
    }
    
    /// Resumes the card.
    /// - Parameters:
    ///   - cardId: Card ID.
    ///   - callback: Callback.
    public func resumeCard(cardId: String, callback: @escaping (D1Error?) -> Void) {
       // self.getD1Task().updateCard(cardId, cardAction: CardAction.resume, reason: "Resume", completion: callback)
    }
    
    /// Deletes the card.
    /// - Parameters:
    ///   - cardId: Card ID.
    ///   - callback: Callback.
    public func deleteCard(cardId: String, callback: @escaping (D1Error?) -> Void) {
       // self.getD1Task().updateCard(cardId, cardAction: CardAction.delete, reason: "Delete", completion: callback)
    }
    
    /// Checks if card is digitized.
    /// - Parameters:
    ///   - cardId: Card ID.
    ///   - callback: Callback.
    public func getDigitizationState(cardId: String, callback: @escaping (CardDigitizationResult?, D1Error?) -> Void) {
        self.getD1Task().cardDigitizationState(cardId, completion: callback)
    }
    
    /// Digitizes the card.
    /// - Parameters:
    ///   - cardid: Card ID.
    ///   - callback: Callback.
    public func digitizeCard(cardid: String, callback: @escaping (D1Error?) -> Void) {
        let viewController: UIViewController = UIViewController()
        getD1Task().addDigitalCardToOEM(cardid, viewController: viewController, completion: callback)
    }
    
    /// Checks if the D1 SDK is configured.
    /// - Returns: True if configured, else false
    public func isConfigured() -> Bool {
        return self.d1Task != nil
    }
    
    /// Retrieves the list of digital cards.
    /// - Parameters:
    ///   - cardId: Card ID.
    ///   - callback: Callback.
    public func getDigitalCardList(cardId: String, callback: @escaping ([DigitalCard]?, D1Error?) -> Void) {
        self.getD1Task().digitalCardList(cardId, completion: callback)
    }
    
    /// Suspends the digital card.
    /// - Parameters:
    ///   - cardId:  Card ID.
    ///   - digitalCard: Digital card.
    ///   - callback: Callback.
    public func suspendDigitalCard(cardId: String, digitalCard: DigitalCard, callback: @escaping (Bool, D1Error?) -> Void) {
        self.getD1Task().updateDigitalCard(cardId, digitalCard: digitalCard, action: .suspend, completion: callback)
    }
    
    /// Resumes the digital card.
    /// - Parameters:
    ///   - cardId: Card ID.
    ///   - digitalCard: Digital card.
    ///   - callback: Callback.
    public func resumedDigitalCard(cardId: String, digitalCard: DigitalCard, callback: @escaping (Bool, D1Error?) -> Void) {
        self.getD1Task().updateDigitalCard(cardId, digitalCard: digitalCard, action: .resume, completion: callback)
    }
    
    /// Deletes the digital card.
    /// - Parameters:
    ///   - cardId: Card ID.
    ///   - digitalCard: Digital card.
    ///   - callback: Callback.
    public func deleteDigitalCard(cardId: String, digitalCard: DigitalCard, callback: @escaping (Bool, D1Error?) -> Void) {
        self.getD1Task().updateDigitalCard(cardId, digitalCard: digitalCard, action: .delete, completion: callback)
    }
    
    /// Activates the physical card.
    /// - Parameters:
    ///   - cardId: Card ID.
    ///   - secureTextField: Secure Text Field for code entry.
    ///   - callback: Callback.
    public func activatePhysicalCard(cardId: String, secureTextField: D1SecureTextField, callback: @escaping (D1Error?) -> Void) {
        let entryUi: EntryUI = EntryUI(entryTextField: secureTextField)
        self.getD1Task().activatePhysicalCard(cardId, entryUI: entryUi, completion: callback)
    }
    
    // Activate digital for inApp
    public func activateDigitalCard(cardId: String, callback: @escaping (D1Error?) -> Void) {
        self.getD1Task().activateDigitalCard(cardId, completion: callback)
    }

    /// Gets the physical card activation method.
    /// - Parameters:
    ///   - cardId: Card ID.
    ///   - callback: Callback.
    public func getActivationMehod(cardId: String, callback: @escaping (CardActivationMethod?, D1Error?) -> Void) {
        self.getD1Task().cardActivationMethod(cardId, completion: callback)
    }
    
    /// Retrieves the physical card PIN.
    /// - Parameters:
    ///   - cardId: Card ID.
    ///   - textField: Text Field to dispaly the PIN.
    ///   - callback: Callback
    public func getPin(cardId: String, textField: D1PINDisplayTextField, callback: @escaping (D1Error?) -> Void ) {
        let cardPinUI: CardPINUI = CardPINUI(pinTextField: textField)
        self.getD1Task().displayPhysicalCardPIN(cardId, cardPINUI: cardPinUI, completion: callback)
    }
    
    /// Retrieves the SDK versions.
    /// - Returns: SDK versions
    func getLibVersions() -> String {
        return D1Task.getSDKVersions().description
    }

    /// Dispatch to main thread with delay - used for development.
    /// - Parameter block: Code block to dispatch.
    private func mockDispatch(block: @escaping () -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            sleep(1)
            DispatchQueue.main.async {
                block()
            }
        }
    }
    
    /// Retrieves the D1Task instance.
    /// - Returns: D1Task instance.
    private func getD1Task() -> D1Task {
        if let d1Task = d1Task {
            return d1Task
        } else {
            fatalError("Need to configure D1 SDK first.")
        }
    }
     
    /// Configures RASP.
    private func configureRasp() {
        // removed
    }
}
